File: python/count_good_nodes.py
class Solution(object):
    # O(n) time | O(n) space
    def goodNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        def dfs(node, pathMax):
            if not node:
                return 0
            count = 0
            if node.val >= pathMax:
                count += 1
            pathMax = max(pathMax, node.val)
            return count + dfs(node.left, pathMax) + dfs(node.right, pathMax)

        return dfs(root, float("-infinity"))


class TreeNode(object):
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __repr__(self):
        result = str(self.val)
        result += str(self.left) if self.left else ''
        result += str(self.right) if self.right else ''
        return result
    

root = TreeNode(3)
root.left = TreeNode(1)
root.right = TreeNode(4)
root.left.left = TreeNode(3)
root.right.left = TreeNode(1)
root.right.right = TreeNode(5)

solution = Solution()
print(solution.goodNodes(root))

File: src/main/java/august_2025/CountGoodNodes.java
package august_2025;

public class CountGoodNodes {

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(1);
        root.right = new TreeNode(4);
        root.left.left = new TreeNode(3);
        root.right.left = new TreeNode(1);
        root.right.right = new TreeNode(5);

        CountGoodNodes countGoodNodes = new CountGoodNodes();
        int result = countGoodNodes.goodNodes(root);
        System.out.println(result);
    }

    // O(n) time | O(n) space
    public int goodNodes(TreeNode root) {
        return helper(root, root.val);
    }

    private int helper(TreeNode node, int max) {
        if (node == null) {
            return 0;
        }
        max = Math.max(node.val, max);
        int count = 0;
        count += helper(node.left, max);
        count += helper(node.right, max);
        if (node.val >= max) {
            count++;
        }
        return count;
    }

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        public TreeNode(int val) {
            this.val = val;
        }
    }

}
